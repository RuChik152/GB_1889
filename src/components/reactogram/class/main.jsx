import React, { Component } from 'react';
import { Chat } from './chat';
import { Input } from './input';

export class Reactogram extends Component {
  state = {
    text: '',
    author: '',
    activate: true,
    time: '',
    msg: [],
  };

  componentDidMount() {
    this.setState({ author: this.props.data });
  }

  handelChange = (e) => {
    let tag = e.target;
    if (tag.className === 'text__input') {
      this.setState({ text: tag.value });
    }
    this.setState({ time: this.createCurrentTime() });
  };

  actionClick = () => {
    let obj = {
      msg: this.state.text,
      author: this.state.author,
      time: this.state.time,
    };
    this.setState({ msg: [...this.state.msg, obj] });
    this.setState({ activate: !this.state.activate });
  };

  createCurrentTime = () => {
    let time = new Date();
    return `${time.getHours()} : ${time.getMinutes()} : ${time.getSeconds()}`;
  };

  updateState = (obj) => {
    this.setState({ msg: [...this.state.msg, obj] });
    this.setState({ time: '' });
    this.setState({ text: '' });
  };

  bot = () => {
    console.log('11');
    let lengthMsg = this.state.msg.length;
    if (lengthMsg === 0) {
      let obj = {
        msg: 'Привет, это бот психологической помощи Вася Вася. Вы можете использовать кодовые фразы для взаимодействия с ним =>  #Слово #Время #Поиск',
        author: 'BothFather',
        time: `${this.createCurrentTime()}`,
      };
      this.updateState(obj);
    } else if (this.state.text === '#Слово') {
      let obj = {
        msg: (
          <>
            <li>
              <a
                href="https://slovardalja.net/"
                target="_blank"
                rel="noreferrer"
              >
                Словарь Даля
              </a>
            </li>
            <li>
              <a
                href="https://stih.su/pushkin/"
                target="_blank"
                rel="noreferrer"
              >
                Стихи Пушкина
              </a>
            </li>
          </>
        ),
        author: 'BothFather',
        time: `${this.createCurrentTime()}`,
      };
      this.updateState(obj);
    } else if (this.state.text === '#Время') {
      let obj = {
        msg: (
          <>
            <li>
              <a href="https://time.is/ru/" target="_blank" rel="noreferrer">
                Time
              </a>
            </li>
            <li>
              <a href="https://time100.ru/" target="_blank" rel="noreferrer">
                Time 100
              </a>
            </li>
          </>
        ),
        author: 'BothFather',
        time: `${this.createCurrentTime()}`,
      };
      this.updateState(obj);
    } else if (this.state.text === '#Поиск') {
      let obj = {
        msg: (
          <>
            <li>
              <a href="https://ya.ru/" target="_blank" rel="noreferrer">
                Time
              </a>
            </li>
            <li>
              <a
                href="https://duckduckgo.com/"
                target="_blank"
                rel="noreferrer"
              >
                DuckDuckGo
              </a>
            </li>
            <li>
              <a
                href="https://www.dogpile.com/"
                target="_blank"
                rel="noreferrer"
              >
                DogPile
              </a>
            </li>
          </>
        ),
        author: 'BothFather',
        time: `${this.createCurrentTime()}`,
      };
      this.updateState(obj);
    } else {
      let name = this.state.msg.slice(-1);
      const msgBot = [
        'Сообщение от пользователя',
        'Message from user(англ.)',
        'Message de lutilisateur(франц.)',
        'Nachricht vom Benutzer(немц.)',
        '\n' + '用戶留言(кит.)',
      ];
      let template = '$wel__template$ $nik__name$.';
      let positionMsgBot = parseInt(Math.random() * msgBot.length);
      let regWel = /\$wel__template\$/gi;
      let newStr = template.replace(regWel, msgBot[positionMsgBot]);
      regWel = /\$nik__name\$/gi;
      newStr = newStr.replace(regWel, name[0].author);
      let obj = {
        msg: `${newStr}`,
        author: 'BothFather',
        time: `${this.createCurrentTime()}`,
      };
      this.updateState(obj);
    }
  };

  render() {
    return (
      <div className="home">
        <Chat
          message={this.state.msg}
          bot={this.bot}
          activate={this.state.activate}
        />
        <Input
          click={this.actionClick}
          text={this.state.text}
          author={this.state.author}
          change={this.handelChange}
        />
      </div>
    );
  }
}
