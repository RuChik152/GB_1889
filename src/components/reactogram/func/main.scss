* {
  padding: 0;
  margin: 0;
}


html {
  background: rgb(223,218,218);
  background: linear-gradient(90deg, rgba(223,218,218,1) 0%, rgba(93,152,195,1) 53%, rgba(216,232,145,1) 100%);
}

.home {
  width: 1140px;
  margin: 3vh auto;
  background: rgb(167,167,167);
  background: linear-gradient(90deg, rgba(167,167,167,1) 0%, rgba(72,119,154,1) 53%, rgba(145,156,97,1) 100%);
  padding: 17px;
}

body {
  margin: 0;
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen',
  'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans', 'Helvetica Neue',
  sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

code {
  font-family: source-code-pro, Menlo, Monaco, Consolas, 'Courier New',
  monospace;
}