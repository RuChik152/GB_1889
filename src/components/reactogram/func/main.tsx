import React, { FC, useState } from 'react';
import { BrowserRouter, Route, Routes } from 'react-router-dom';
import { Home } from './pages/Home';
import { Profile } from './pages/Profile';
import { Header } from './Header';
import { Chats } from './pages/Chats';
import { ChatList } from './ChatList/ChatList';
import { AUTHOR, createCurrentTime } from '../modal/utility';

export interface Chat {
  id: string;
  name: string;
}

const initialChat: Chat[] = [
  {
    id: 'default',
    name: 'chat',
  },
];

const initialMsgs: Msgs = {
  default: [
    {
      id: '1',
      author: AUTHOR.user,
      time: createCurrentTime(),
      msg: 'Hello World',
    },
  ],
};

interface Msg {
  id: string;
  author: string;
  time: string;
  msg: string;
}

interface Msgs {
  [key: string]: Msg[];
}

export const Reactogram: FC = () => {
  const [chatlist, setChatlist] = useState<Chat[]>(initialChat);
  const [msg, setMsg] = useState<Msgs>(initialMsgs);

  const addChatList = (chat: Chat) => {
    setChatlist([...chatlist, chat]);
    setMsg({ ...msg, [chat.id]: [] });
  };

  const removeChatList = (e: { target: { dataset: { id: string } } }) => {
    const idxList: number = chatlist.findIndex(
      (item) => item.id == e.target.dataset.id
    );
    setChatlist([...delList(chatlist, e.target.dataset.id)]);
    setMsg({ ...delMsg(msg, e.target.dataset.id) });
  };

  const delList = (arr: Chat[], idx: string) => {
    return arr.filter( (elem) => elem.id !== idx);
  };

  const delMsg = (obj: Msgs, idx: string) => {
    const arr = { ...obj };
    delete arr[idx];
    return arr;
  };

  return (
    <BrowserRouter>
      <Routes>
        <Route path="/" element={<Header />}>
          <Route index element={<Home />} />
          <Route path="profile" element={<Profile />} />
          <Route path="chats">
            <Route
              index
              element={
                <ChatList
                  chatlist={chatlist}
                  addChatList={addChatList}
                  removeChatList={removeChatList}
                />
              }
            />
            <Route
              path=":chaiId"
              element={
                <Chats
                  setMsg={setMsg}
                  msg={msg}
                  chatlist={chatlist}
                  addChatList={addChatList}
                  removeChatList={removeChatList}
                />
              }
            />
          </Route>
        </Route>
        <Route path="*" element={<h2>404</h2>} />
      </Routes>
    </BrowserRouter>
  );
};
