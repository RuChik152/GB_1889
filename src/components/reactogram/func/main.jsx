import React, { useEffect, useState } from 'react';
import { Chat } from './Chat/chat';
import { Input } from './Input/input';
import './main.scss'

export const Reactogram = ({ data }) => {
  const [text, setText] = useState('');
  const [time, setTime] = useState('');
  const [msg, setMsg] = useState([]);

  useEffect(() => {
    setTimeout(bot, 1500);
    //eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    if (msg.length !== 0) {
      if (msg[msg.length - 1].author !== 'BothFather') {
        const timeout = setTimeout(bot, 1500);
        return () => {
          clearTimeout(timeout);
        };
      }
    }
    //eslint-disable-next-line react-hooks/exhaustive-deps
  }, [msg]);

  const handelChange = (e) => {
    let tag = e.target;
    if (tag.className === 'text__input') {
      setText(tag.value);
    }
    setTime(createCurrentTime());
  };

  const actionClick = () => {
    let obj = { msg: text, author: data, time: time };
    setMsg([...msg, obj]);
  };

  const createCurrentTime = () => {
    let time = new Date();
    return `${time.getHours()} : ${time.getMinutes()} : ${time.getSeconds()}`;
  };

  const bot = () => {
    let lengthMsg = msg.length;
    if (lengthMsg === 0) {
      let obj = {
        msg: 'Привет, это бот психологической помощи Вася Вася. Вы можете использовать кодовые фразы для взаимодействия с ним =>  #Слово #Время #Поиск',
        author: 'BothFather',
        time: `${createCurrentTime()}`,
      };
      setMsg([...msg, obj]);
      setTime('');
      setText('');
    }

    if (text === '#Слово') {
      let obj = {
        msg: (
          <>
            <li>
              <a
                href="https://slovardalja.net/"
                target="_blank"
                rel="noreferrer"
              >
                Словарь Даля
              </a>
            </li>
            <li>
              <a
                href="https://stih.su/pushkin/"
                target="_blank"
                rel="noreferrer"
              >
                Стихи Пушкина
              </a>
            </li>
          </>
        ),
        author: 'BothFather',
        time: `${createCurrentTime()}`,
      };
      setMsg([...msg, obj]);
      setTime('');
      setText('');
    } else if (text === '#Время') {
      let obj = {
        msg: (
          <>
            <li>
              <a href="https://time.is/ru/" target="_blank" rel="noreferrer">
                Time
              </a>
            </li>
            <li>
              <a href="https://time100.ru/" target="_blank" rel="noreferrer">
                Time 100
              </a>
            </li>
          </>
        ),
        author: 'BothFather',
        time: `${createCurrentTime()}`,
      };
      setMsg([...msg, obj]);
      setTime('');
      setText('');
    } else if (text === '#Поиск') {
      let obj = {
        msg: (
          <>
            <li>
              <a href="https://ya.ru/" target="_blank" rel="noreferrer">
                Time
              </a>
            </li>
            <li>
              <a
                href="https://duckduckgo.com/"
                target="_blank"
                rel="noreferrer"
              >
                DuckDuckGo
              </a>
            </li>
            <li>
              <a
                href="https://www.dogpile.com/"
                target="_blank"
                rel="noreferrer"
              >
                DogPile
              </a>
            </li>
          </>
        ),
        author: 'BothFather',
        time: `${createCurrentTime()}`,
      };
      setMsg([...msg, obj]);
      setTime('');
      setText('');
    } else {
      let name = msg.slice(-1);
      const msgBot = [
        'Сообщение от пользователя',
        'Message from user(англ.)',
        'Nachricht vom Benutzer(немц.)',
        '\n' + '用戶留言(кит.)',
      ];
      let template = '$wel__template$ $nik__name$.';
      let positionMsgBot = parseInt(Math.random() * msgBot.length);
      let regWel = /\$wel__template\$/gi;
      let newStr = template.replace(regWel, msgBot[positionMsgBot]);
      regWel = /\$nik__name\$/gi;
      newStr = newStr.replace(regWel, name[0].author);
      let obj = {
        msg: `${newStr}`,
        author: 'BothFather',
        time: `${createCurrentTime()}`,
      };
      setMsg([...msg, obj]);
      setTime('');
      setText('');
    }
  };

  return (
    <div className="home" data-testid="home-test-id">
      <Chat message={msg} />
      <Input
        click={actionClick}
        text={text}
        author={data}
        change={handelChange}
      />
    </div>
  );
};
