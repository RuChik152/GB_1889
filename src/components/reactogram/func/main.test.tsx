import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import { Reactogram } from './main';

describe('Main', () => {
  it('Render', () => {
    render(<Reactogram data={'TEST_NAME'} />);
  });

  it('Render input and button', () => {
    const { getByPlaceholderText, getByTestId } = render(<Reactogram data={'TEST_NAME'} />);
    const inputText = getByPlaceholderText('Ваше сообщение');
    const btn = getByTestId('test-id');
    expect(inputText).toBeInTheDocument();
    expect(btn).toBeInTheDocument();
  });

  it('render chat', () => {
    const { getByTestId } = render(<Reactogram data={'TEST_NAME'} />);
    const el = getByTestId('chat-test');
    expect(el).toBeInTheDocument();
  });

  it('Snapshot', () => {
    const { asFragment } = render(<Reactogram data={'TEST_NAME'} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
