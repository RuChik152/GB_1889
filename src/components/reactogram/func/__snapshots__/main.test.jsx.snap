// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main Snapshot 1`] = `
<DocumentFragment>
  <div
    class="home"
    data-testid="home-test-id"
  >
    <div
      class="chat"
      data-testid="chat-test"
    >
      <div
        class="chat__area"
      />
    </div>
    <div
      class="text"
    >
      <input
        class="text__input"
        name="msg"
        placeholder="Ваше сообщение"
        type="text"
        value=""
      />
      <br />
      <button
        class="text__btn"
        data-testid="test-id"
      >
        Отправить
      </button>
    </div>
  </div>
</DocumentFragment>
`;
